### DISCUSSION TITLE
anything vs any vs anyInstanceOf

### DISCUSSION THREAD

**Discussion by Janglinator** [mockative/mockative#15] [Q&A] [answered]
What is the difference between these 3 matchers?

```
inline fun <reified T> anything(): AnythingMatcher<T> {
    return AnythingMatcher()
}

inline fun <reified T> any(): AnyMatcher<T> {
    return AnyMatcher(T::class)
}

fun <R : Any, T : R> anyInstanceOf(type: KClass<T>): AnyMatcher<R> {
    return AnyMatcher(type)
}
```

It seems `anything` is exactly that, and `anyInstanceOf` is intuitive, but what does `any` imply?  I can't seem to write a test that succeeds with it.

All the examples in the README are written using `any`, so I feel I should know what it does.

Looking through the source code, it seems like any is nearly equivalent to anyInstanceOf, but with generics, but i still can't get it to work.

**Comment by Nillerr** [comment, marked as answer]
To address your first question:
- `anything()` matches _anything_, regardless of the type of the argument
- `any()` matches _any_ value, validating that the type is a specific type.
- `anyInstanceOf(KClass<T>)` matches any instance of a specific type, which may be useful when using sealed classes or other types of inheritance trees.

`any()` and `anyInstanceOf(KClass<T>)` both does the same, but `anyInstanceOf(KClass<T>)` allows you to trivially specify a more specific type to match. You'll rarely if ever want to use `anything()`. We simply included it because it was trivial, so in case someone needs it it's there.

As to why you can't get it working, it would be very helpful to see a snippet of what you're trying to do with the matchers. The following snippet produces a successful test:

```kotlin
    @Test
    fun givenSetupOfSuspendingCommandWithAnyArgument_whenCallingCommand_thenMockIsCalled() = runBlockingTest {
        // given
        val id = "0efb1b3b-f1b2-41f8-a1d8-368027cc86ee"
        val repository = Repository(id, "Mockative")

        given(github).suspendFunction(github::repository)
            .whenInvokedWith(any())
            .thenReturn(repository)

        // when
        service.repository(id)

        // then
        verify(github).suspendFunction(github::repository)
            .with(any())
            .wasInvoked(exactly = once)
    }
```

Please bear in mind that the matcher API only works with the function reference API:

```kotlin
given(mock).function(mock::myFunction)
  .whenInvokedWith(any())
  .then { ... }

given(mock).suspendFunction(mock::mySuspendFunction)
  .whenInvokedWith(any())
  .then { ... }
```

It does **not** work with the invocation API:

```kotlin
given(mock).invocation { myFunction("myString") }
  .then { ... }
```
